### PR TITLE
[YUNIKORN-982] placeholder size checks

### PR DESCRIPTION
**Pull request by wilfred-s**
### What is this PR for?
Check placeholder during allocation cycle if placeholder is smaller than the real allocation the placeholder(s) get removed.
Check placeholder during replacement if placeholder is larger than the real allocation the node and queue usage is adjusted.
Additional unit test:
partition: rewrite of placeholder and real alloc tests
node: new test for delta processing on replace
application: test fixes only

### What type of PR is it?
* [X] - Bug Fix

### What is the Jira issue?
https://issues.apache.org/jira/browse/YUNIKORN-982

### How should this be tested?
New unit tests added for all use cases

### CODE REVIEW

File: pkg/scheduler/partition.go [hunk 1]
```diff
@@ -704,6 +704,28 @@ func (pc *PartitionContext) removeNodeAllocations(node *objects.Node) ([]*object
...
 					// ignore the return as that is the same as alloc, the alloc is gone after this call
 					_ = app.ReplaceAllocation(allocID)
+					// we need to check the resources equality
+					delta := resources.Sub(release.AllocatedResource, alloc.AllocatedResource)
+					// Any negative value in the delta means that at least one of the requested resource in the
+					// placeholder is larger than the real allocation. The nodes are correct the queue needs adjusting.
+					// The reverse case is handled during allocation.
+					if delta.HasNegativeValue() {
+						// this looks incorrect but the delta is negative and the result will be a real decrease
+						err := app.GetQueue().IncAllocatedResource(delta, false)
```
Comment (manirajv06) [inline pkg/scheduler/partition.go:714]: Would this new queue adjustment cause any issues in conjunction with below adjustment in place? app.GetQueue().DecAllocatedResource(alloc.AllocatedResource)

Can we double check? May be unit test to removeNode with replacement on different node ensures this?

Other than this, overall changes looks good.

**Review by manirajv06 — CHANGES_REQUESTED** [review]

File: pkg/scheduler/objects/application.go [hunk 2]
```diff
@@ -827,13 +827,41 @@ func (sa *Application) tryPlaceholderAllocate(nodeIterator func() NodeIterator,
 			if ph.released || request.taskGroupName != ph.taskGroupName {
 				continue
 			}
+			// before we check anything we need to check the resources equality
+			delta := resources.Sub(ph.AllocatedResource, request.AllocatedResource)
```
Comment (chenya-zhang) [inline pkg/scheduler/objects/application.go:831]: Maybe we could replace "ph" (placeholder) with "hold" (v.s. "request") to make our codebase easier to understand. :) But if it appears in too many places, it's okay.

File: pkg/scheduler/objects/application.go [hunk 3]
```diff
@@ -827,13 +827,41 @@ func (sa *Application) tryPlaceholderAllocate(nodeIterator func() NodeIterator,
...
+			// Any negative value in the delta means that at least one of the requested resource in the real
+			// allocation is larger than the placeholder. We need to cancel the placeholder and check the next
+			// placeholder. This should trigger a cleanup of all the placeholder as a task group is always one size.
```
Comment (chenya-zhang) [inline pkg/scheduler/objects/application.go:834]: > This should trigger a cleanup of all the placeholder as a task group is always one size.

It sounds to me that all other placeholders will be removed (= cleaned up)?

File: pkg/scheduler/objects/application.go [hunk 4]
```diff
@@ -827,13 +827,41 @@ func (sa *Application) tryPlaceholderAllocate(nodeIterator func() NodeIterator,
...
+			if delta.HasNegativeValue() {
+				log.Logger().Warn("releasing placeholder: real allocation is larger than placeholder",
+					zap.String("requested resource", request.AllocatedResource.String()),
+					zap.String("placeholderID", ph.UUID),
+					zap.String("placeholder resource", ph.AllocatedResource.String()))
+				// release the placeholder and tell the RM
+				ph.released = true
+				sa.notifyRMAllocationReleased(sa.rmID, []*Allocation{ph}, si.TerminationType_TIMEOUT, "cancel placeholder: resource incompatible")
```
Comment (chenya-zhang) [inline pkg/scheduler/objects/application.go:842]: > cancel placeholder: resource incompatible

Will it be easier to understand if we say "requested resource larger than placeholder" directly? "resource incompatible" may also refer to resource type.

File: pkg/scheduler/objects/application.go (hunk 3 above)
Comment (pbacsko) [reply pkg/scheduler/objects/application.go:834]: I would rephrase it as "task group has a size of one", sounds more correct.

File: pkg/scheduler/objects/application.go (hunk 4 above)
Comment (pbacsko) [inline pkg/scheduler/objects/application.go:842]: `si.TerminationType_TIMEOUT` - can this be changed to something else or would we break API with that? Something like `si.TerminationType_RESOURCE_MISMATCH`.

File: pkg/scheduler/objects/node_test.go [hunk 5]
```diff
@@ -488,6 +488,40 @@ func TestRemoveAllocation(t *testing.T) {
...
+	allocID := "real-1"
+	piece := resources.NewResourceFromMap(map[string]resources.Quantity{"first": 25, "second": 50})
+	alloc := newAllocation(appID1, allocID, nodeID1, "queue-1", piece)
+	// calculate the delta: new allocation resource - placeholder (should be negative!)
+	delta := resources.Sub(piece, half)
+	assert.Assert(t, delta.HasNegativeValue(), "expected negative values in delta")
+	// swap and check the calculation
+	node.ReplaceAllocation(phID, alloc, delta)
+	if node.GetAllocation(allocID) == nil {
+		t.Fatal("failed to replace allocation: allocation not returned")
```
Comment (pbacsko) [inline pkg/scheduler/objects/node_test.go:514]: Nit: can't this be an assert as well?

File: pkg/scheduler/objects/node_test.go [hunk 6]
```diff
@@ -488,6 +488,40 @@ func TestRemoveAllocation(t *testing.T) {
...
+	}
+	assert.Assert(t, resources.Equals(node.GetAllocatedResource(), piece), "allocated resource not set correctly %v got %v", piece, node.GetAllocatedResource())
+
+	// clean up all should be zero
+	if node.RemoveAllocation(allocID) == nil {
+		t.Error("allocation should have been removed but was not")
```
Comment (pbacsko) [inline pkg/scheduler/objects/node_test.go:520]: Same here

File: pkg/scheduler/objects/node.go [hunk 7]
```diff
@@ -316,15 +316,17 @@ func (sn *Node) AddAllocation(alloc *Allocation) bool {
 	return false
 }
 
-// Replace the paceholder allocation on the node. No usage changes as the placeholder must
-// be the same size as the real allocation.
-func (sn *Node) ReplaceAllocation(uuid string, replace *Allocation) {
+// ReplaceAllocation replaces the placeholder with the real allocation on the node.
+// The delta passed in is the difference in resource usage between placeholder and real allocation.
+// It should always be a negative value (really a decreased usage).
```
Comment (pbacsko) [inline pkg/scheduler/objects/node.go:321]: If this is a strong requirement, then it might make sense to at least log a warning if this expectation (`delta.HasNegativeValue() == true`) is not met. This can lead to bugs that are hard to figure out.

File: pkg/scheduler/objects/application.go (hunk 3 above)
Comment (wilfred-s) [reply pkg/scheduler/objects/application.go:834]: Updated comment to be clearer

File: pkg/scheduler/objects/application.go (hunk 4 above)
Comment (wilfred-s) [reply pkg/scheduler/objects/application.go:842]: I did not introduce a new state as the processing for both would be exactly the same in both the shim and the core. It would also require a new SI and a new shim to make the same changes work.

Comment (wilfred-s) [reply pkg/scheduler/objects/application.go:842]: Correct because the resource could be incompatible. The real allocation has a reference to resource type GPU with a value of 1, but the placeholder does not which results in GPU becoming negative in the `delta`. The same real allocation could also ask for memory 2GB while the placeholder specifies 4GB. The "incompatible resource" covers these mixed cases better than just the larger than message.
The event generated,  which is what the end user sees, has all the detail needed.

File: pkg/scheduler/objects/node.go (hunk 7 above)
Comment (wilfred-s) [reply pkg/scheduler/objects/node.go:321]: added log entry

File: pkg/scheduler/objects/node_test.go (hunk 5 above)
Comment (wilfred-s) [reply pkg/scheduler/objects/node_test.go:514]: We should file a follow up for that, the node testing code has a large number of places where it should  be using asserts.

Nil object checks do look a bit weird in the assert, might not improve readability:
```
assert.Assert(t, node.GetAllocation(allocID) != nil, "failed to replace allocation: allocation not returned")
```

File: pkg/scheduler/partition.go (hunk 1 above)
Comment (wilfred-s) [reply pkg/scheduler/partition.go:714]: The update of the queue is enclosed in a check for nodeID difference. The last line of that block is a continue call. when we get into the block we can never reach line 770 which does the queue update for the other cases.

File: pkg/scheduler/objects/application.go (hunk 2 above)
Comment (wilfred-s) [reply pkg/scheduler/objects/application.go:831]: too big a change

**Review by pbacsko — APPROVED** [review]
LGTM +1

**Review by craigcondit — APPROVED** [review]
+1. LGTM.

**Review by chenya-zhang — APPROVED** [review]
+1 LGTM, thanks @wilfred-s!
